### PR TITLE
Compile-time signed_saturate bits parameter

### PR DESCRIPTION
**Pull request by stellar-aria**
This PR is to fix a build issue I found.

## Background
Unlike all of the other inline-assembly functions that are provided by `functions.h`, signed_saturate has one crucial difference:

The ARMv7A `ssat` instruction takes an immediate, not a register, for its second operand.

## The Problem
At sufficiently high optimization levels, this doesn't present a problem, as compile-time constants are combined and reduced down to a single immediate value before reaching the assembler. However, this is technically just hiding the problem that the signed_saturate function is taking what is potentially a run-time variable and trying to insert it into assembly code.

The assembler is, expectedly, not very happy about this when it tries to happen:
```
DelugeFirmware/src/functions.h:108:9: warning: 'asm' operand 1 probably does not match constraints
  108 |         asm("ssat %0, %1, %2" : "=r"(out) : "I"(bits), "r"(val));
      |         ^~~
DelugeFirmware/src/functions.h:108:9: error: impossible constraint in 'asm'
```
If the constraint is expanded to 'X' (any), we can see the source of the issue at the assembly stage:
```
Error: constant expression required -- `ssat r3,[fp,#-9],r3'
```
Which is, of course, that the compiler inserted a reference to a stack variable and the assembler has no idea what to do with it.

## The Solution
Thankfully, C++ already has facilities for strongly enforcing compile-time code evaluation of constants: templates.

By simply moving 'bits' from a function parameter to a template parameter, the compiler does this calculation before the generated assembly is passed off and everything is kept happy.

This does unfortunately trickle down to any functions that pass their own parameters to `signed_saturate`, but that's how compile-time constraints are meant to work in the first place.

### CODE REVIEW

**Comment by litui** [comment]
!!! Thank you for this!  I think this is also related to the Assembler compile issues I was having in arm-none-eabi-gcc 12.2. I'll take this for a spin and see how it goes!

**Comment by litui** [comment]
Functionally, this seems to be solid. I built successfully under both arm-none-eabi-gcc 9.2.1 and 12.2 and tested both in debug builds on the Deluge 7seg under e2studio in Windows.

File sizes are slightly smaller under 12.2 which is nice to see. Great PR!

Just the clang formatting check needs to pass before I can give the merge go-ahead.

**Comment by stellar-aria** [comment]
Patched!

File: src/functions.cpp
Comment (bobtwinkles) [inline src/functions.cpp:216]: nit: move the comment above so we get slightly better wrapping behavior here

**Comment by bobtwinkles** [comment]
Minor nit that you absolutely don't have to address if it's annoying, but otherwise this looks like a nice safety improvement.

File: src/functions.cpp
Comment (stellar-aria) [reply src/functions.cpp:216]: yeah that's fair

**Review by litui — APPROVED** [review]

**Pull request merged**
